src/centrality/truss.cpp

```diff
@@ -0,0 +1,271 @@
+  IGRAPH_FINALLY_CLEAN(1);
+
+  // Compute the truss of the edges.
+  IGRAPH_CHECK(igraph_vector_int_init(truss, igraph_ecount(graph)));
+  IGRAPH_CHECK(igraph_i_trussness(graph, &support, truss));
+  igraph_vector_int_destroy(&support);
+  IGRAPH_FINALLY_CLEAN(1);
+
+  return IGRAPH_SUCCESS;
+}
+
+// Unpack the triangles as a vector of vertices to be a vector of edges.
+// So, instead of the triangle specified as vertices [1, 2, 3], return the
+// edges as [1, 2, 1, 3, 2, 3] so that the support can be computed.
+void igraph_truss_i_unpack(const igraph_vector_int_t *tri, igraph_vector_t *unpacked_tri) {
+  long int i, j;
+  for (i = 0, j = 0; i < igraph_vector_int_size(tri); i += 3, j += 6){
+    VECTOR(*unpacked_tri)[j]   = VECTOR(*unpacked_tri)[j+2] = VECTOR(*tri)[i];
+    VECTOR(*unpacked_tri)[j+1] = VECTOR(*unpacked_tri)[j+4] = VECTOR(*tri)[i+1];
+    VECTOR(*unpacked_tri)[j+3] = VECTOR(*unpacked_tri)[j+5] = VECTOR(*tri)[i+2];
+  }
+}
+
+// Compute the edge support, i.e. number of triangles each edge occurs in.
+// Time complexity: O(m), where m is the number of edges listed in eid.
+void igraph_truss_i_compute_support(const igraph_vector_t *eid, igraph_vector_int_t *support) {
+  long int m = igraph_vector_size(eid);
+  for (long int i = 0; i < m; ++i){
+    VECTOR(*support)[(long int) VECTOR(*eid)[i]] += 1;
+  }
+}
+
+
+/**
+ * \function trussness
+ * \brief Find the "trussness" for every edge in the graph. This
+ * indicates the highest k-truss that the edge occurs in.
+ *
+ * This may differ from the literature in that the value returned is 2 less.
+ *
+ * </para><para>
+ * A k-truss is a subgraph in which every edge occurs in at least k triangles in the subgraph.
+ *
+ * </para><para>
+ * This function returns the highest k for each edge. If you are interested in
+ * a particular k-truss subgraph, you can subset the graph using to those eids
+ * which are >= k because each k-truss is a subgraph of a (k–1)-truss
+ * (thus to get * all 4-trusses, take k >= 4 because the 5-trusses, 6-trusses, etc need to be included).
+ *
+ * </para><para>The current implementation of this function iteratively
+ * decrements support of each edge using O(|E|) space and O(|E|^1.5) time.
+ *
+ * \param graph The input graph.
+ * \param support The support of each edge, specified as a vector where the
+ * indices stand for the edges, and support is (initially!) the number of
+ * triangles the edge occurs in. This is used as a workspace and is changed,
+ * then destroyed.
+ * \param truss This will indicate the highest k-truss each edge occurs in.
+ * \return Error code.
+ *
+ *
+ * Time complexity: It should be O(|E|^1.5). See Algorithm 2 in:
+ * Wang, Jia, and James Cheng. "Truss decomposition in massive networks."
+ * Proceedings of the VLDB Endowment 5.9 (2012): 812-823.
+ *
+ *
+ * \example test/unit/truss.cpp FIXME: can we make a C file instead?
+ */
+int igraph_i_trussness(const igraph_t *graph, igraph_vector_int_t *support,
```

Comment:
This function is documented, but it is now internal. @alexperrone is this meant as a public function?

Reply:
probably not, it's where the heavy lifting happens after support has been computed...